icicle/appUtils/msm/msm.cu

```diff
@@ -135,9 +135,9 @@ __global__ void find_cutoff_kernel(unsigned *v, unsigned size, unsigned cutoff,
       result[0] = i+1;
       return;
     }
-    if (i == size - 1) {
-      result[0] = 0;
-    }
+  }
+  if (v[size - 1] > cutoff) {
+    result[0] = size;
```

Comment:
why did you change this? also, now all the threads write to result[0] instead of a single thread..

Reply:
Tbh I don't understand the original version. First of all, `i` is never `size - 1` because the loop terminates at `i<min(start_index+run_length,size-1)`, so your block is never executed, this breaks correctness (for 0;1;-1 distribution for example). And judging from the condition, it is supposed to work when all the nonzero buckets are large, right? In that case, why would `result` be 0 and not `size`? In the case of 0;1;-1 distribution I would like -1 buckets to be treated as large and not as small

Reply:
About threads writing simultaneously - I agree that looks awkward but because the value is the same for all threads, it's not a data race, and I don't think there's a performance penalty as one thread has to diverge to return the result anyway

Reply:
I already answered on slack.. the fix should be i<min(start_index+run_length,size), then the last thread (and only it) always reaches size -1. I put 0 so the regular bucket kernel would run, but I guess putting size is better yeah.

Reply:
In this case we'll be accessing `v[size + 1]` which hypothetically might cause a segfault, and even if the condition `if (i == size - 1)` is put before the main body of the loop to avoid this, you'll still need to add `v[size - 1] > cutoff` condition to it, otherwise you're creating a data race, no?
And checking the condition on each iteration of the loop is less efficient than just once in the end too...